processing/src/main/java/org/apache/druid/segment/serde/cell/DeserializingIOIterator.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.serde.cell;
+
+import com.google.common.base.Preconditions;
+
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.NoSuchElementException;
+
+public class DeserializingIOIterator<T> implements IOIterator<T>
```

Comment:
see comment above; it actually makes the BufferStore *more* complicated because of the need for a template-and-hook to process the row. You don't really buy much and it heads down the path of making an abstract base class with all kinds of hooks for doing 2-pass encoding (since it's during the first pass you collect info--min/max numbers, hash tables for dictionary encoding, etc)

I'd prefer to wait and see how the double-sketch looks (I have a good idea, being this it will be the 3rd implementation I've done of column storage)

Reply:
ugh, i'm slow, nevermind that. i see how we can pull out just the storage and retrieval as pure delegation.

lemme see what it looks like. it's more useful to have a class that allows you to store objects than just the DeserializingIOIterator..